### PR TITLE
MDEV-17048 Inconsistency voting support

### PR DESCRIPTION
**Pull request by ayurchen**
* Collect and pass apply error data to provider
 * Rollback failed transaction and continue operation if provider returns
   SUCCESS
 * MTR tests for inconsistency voting

### CODE REVIEW

**Comment by janlindstrom** [comment]
In my laptop, I saw only this:

galera_sr.galera_sr_cc_master 'innodb'   w6 [ fail ]  Found warnings/errors in server log file!
        Test ended at 2019-08-20 12:11:42
line
safe_mutex: Found wrong usage of mutex 'LOCK_wsrep_cluster_config' and 'LOCK_wsrep_server_state'
^ Found warnings in /home/jan/mysql/10.5/mysql-test/var/6/log/mysqld.2.err
ok

 - saving '/home/jan/mysql/10.5/mysql-test/var/6/log/galera_sr.galera_sr_cc_master-innodb/' to '/home/jan/mysql/10.5/mysql-test/var/log/galera_sr.galera_sr_cc_master-innodb/'

**Comment by janlindstrom** [comment]
http://buildbot.askmonty.org/buildbot/grid?category=main&branch=bb-10.5-galera

File: mysql-test/suite/galera_3nodes/t/GCF-363.test [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+--let $wait_condition = SELECT VARIABLE_VALUE = 'OFF' FROM INFORMATION_SCHEMA.GLOBAL_STATUS WHERE VARIABLE_NAME = 'wsrep_ready'
+--source include/wait_condition.inc
+SHOW STATUS LIKE 'wsrep_cluster_status';
+
+--connection node_1
+# Wait until node #3 leaves the cluster
+--let $wait_condition = SELECT VARIABLE_VALUE = 2 FROM INFORMATION_SCHEMA.GLOBAL_STATUS WHERE VARIABLE_NAME = 'wsrep_cluster_size'
+--source include/wait_condition.inc
+
+# need to use another connection to restart the node due to GCF-1000
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera_3nodes/t/GCF-363.test]: True, this issue is long outdated and the node failed for inconsistency better be restarted anyways.

File: mysql-test/suite/galera_3nodes/t/galera-features#119.test [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+--connection node_1
+SET GLOBAL wsrep_provider_options = 'gmcast.isolate=0';
+--source include/galera_wait_ready.inc
+SET GLOBAL wsrep_sync_wait=15;
+DROP TABLE test.t1;
+
+# reconnect node 2, since it is now inconsistent
+--connection node_2
+--source include/kill_galera.inc
+--sleep 1
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera_3nodes/t/galera-features#119.test]: Good catch, thanks!

File: mysql-test/suite/galera/t/GCF-360.test [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+#
+
+--source include/galera_cluster.inc
+
+--let $count = 100
+
+--connect node_3, 127.0.0.1, root, , test, $NODE_MYPORT_3
+--connect node_4, 127.0.0.1, root, , test, $NODE_MYPORT_4
+
+--disable_query_log
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera/t/GCF-360.test]: done

File: mysql-test/suite/galera/t/GCF-360.test [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        --error ER_BAD_TABLE_ERROR
+        --reap
+
+        --dec $count
+}
+
+--enable_query_log
+
+--connection node_1
+SELECT VARIABLE_VALUE = 4 FROM INFORMATION_SCHEMA.GLOBAL_STATUS WHERE VARIABLE_NAME = 'wsrep_cluster_size';
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera/t/GCF-360.test]: done

File: mysql-test/suite/galera/t/galera_vote_rejoin_ddl.test [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+--let $wait_condition = SELECT VARIABLE_VALUE = 3 FROM INFORMATION_SCHEMA.GLOBAL_STATUS WHERE VARIABLE_NAME = 'wsrep_cluster_size';
+--source include/wait_condition.inc
+
+--connection node_1
+--let $wait_condition = SELECT VARIABLE_VALUE = 3 FROM INFORMATION_SCHEMA.GLOBAL_STATUS WHERE VARIABLE_NAME = 'wsrep_cluster_size';
+--source include/wait_condition.inc
+
+# Confirm that all is good
+--connection node_3
+SELECT COUNT(*) = 0 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = 't1';
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera/t/galera_vote_rejoin_ddl.test]: done

File: mysql-test/suite/galera/t/galera_vote_rejoin_ddl.test [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+--echo Starting mysqld
+--source include/start_mysqld.inc
+
+--connection node_1
+# Confirm node #4 has rejoined
+--let $wait_condition = SELECT VARIABLE_VALUE = 4 FROM INFORMATION_SCHEMA.GLOBAL_STATUS WHERE VARIABLE_NAME = 'wsrep_cluster_size';
+--source include/wait_condition.inc
+
+# Confirm that all is good and all nodes have identical data
+SELECT COUNT(*) = 0 FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = 't1';
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera/t/galera_vote_rejoin_ddl.test]: done

File: mysql-test/suite/galera/t/galera_vote_rejoin_dml.test [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+--let $wait_condition = SELECT VARIABLE_VALUE = 3 FROM INFORMATION_SCHEMA.GLOBAL_STATUS WHERE VARIABLE_NAME = 'wsrep_cluster_size';
+--source include/wait_condition.inc
+
+--connection node_1
+--let $wait_condition = SELECT VARIABLE_VALUE = 3 FROM INFORMATION_SCHEMA.GLOBAL_STATUS WHERE VARIABLE_NAME = 'wsrep_cluster_size';
+--source include/wait_condition.inc
+
+# Confirm that all is good
+--connection node_3
+SELECT COUNT(*) = 1 FROM t1 WHERE f2 = 'A';
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera/t/galera_vote_rejoin_dml.test]: done

File: mysql-test/suite/galera_3nodes/t/GCF-376.test [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+LOCK TABLE t1 WRITE;
+
+--connection node_1
+INSERT INTO t1 VALUES (1, 'b');
+
+SET GLOBAL wsrep_provider_options = 'gmcast.isolate=1';
+SET SESSION wsrep_sync_wait=0;
+
+# --sleep 10
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera_3nodes/t/GCF-376.test]: fixed

File: mysql-test/suite/galera_3nodes/t/galera_vote_rejoin_mysqldump.test [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+--source include/wait_condition.inc
+--source include/galera_wait_ready.inc
+SET SESSION wsrep_on=ON;
+
+# restart node so we don't fail on WSREP_START_POSITION internal check
+--source include/restart_mysqld.inc
+--source include/wait_until_connected_again.inc
+
+SHOW CREATE TABLE t1;
+SELECT COUNT(*) = 0 FROM t1;
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera_3nodes/t/galera_vote_rejoin_mysqldump.test]: done

File: mysql-test/suite/galera_3nodes/t/inconsistency_shutdown.test [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+UPDATE t1 SET f2 = 1 WHERE f1 = 1;
+UPDATE t1 SET f2 = 1 WHERE f1 = 2;
+UPDATE t1 SET f2 = 1 WHERE f1 = 3;
+UPDATE t1 SET f2 = 1 WHERE f1 = 4;
+UPDATE t1 SET f2 = 2 WHERE f1 = 4; /* dependent applier */;
+
+# interleave a failing statement
+--connect node_2a, 127.0.0.1, root, , test, $NODE_MYPORT_2
+--connection node_2a
+--send DROP TABLE t2;
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera_3nodes/t/inconsistency_shutdown.test]: fixed using sync point

File: mysql-test/suite/galera_3nodes_sr/t/galera_vote_sr.inc [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+SET AUTOCOMMIT=OFF;
+SET SESSION wsrep_trx_fragment_size = 131070;
+START TRANSACTION;
+INSERT INTO t1 VALUES (1, REPEAT('A', 65535));
+INSERT INTO t1 VALUES (2, REPEAT('A', 65535));
+INSERT INTO t1 VALUES (3, REPEAT('A', 65535));
+INSERT INTO t1 VALUES (4, REPEAT('A', 65535));
+INSERT INTO t1 VALUES (5, REPEAT('A', 65535));
+COMMIT;
+SELECT COUNT(*) = 0 FROM mysql.wsrep_streaming_log;
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera_3nodes_sr/t/galera_vote_sr.inc]: done

File: mysql-test/suite/galera_3nodes_sr/t/galera_vote_sr.inc [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+# Node #2 should have some entries in its SR table post-restart
+SELECT COUNT(*) > 0 FROM mysql.wsrep_streaming_log;
+
+# Commit second SR transaction
+--connection node_1
+COMMIT;
+
+# Confirm that all nodes are identical
+--connection node_1
+SELECT COUNT(*) = 0 FROM mysql.wsrep_streaming_log;
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply mysql-test/suite/galera_3nodes_sr/t/galera_vote_sr.inc]: done

File: sql/wsrep_high_priority_service.cc [hunk 13]
```diff
@@ -358,13 +382,8 @@ int Wsrep_high_priority_service::apply_toi(const wsrep::ws_meta& ws_meta,
...
 
-  int ret= wsrep_apply_events(thd, m_rli, data.data(), data.size());
-  if (ret != 0 || thd->wsrep_has_ignored_error)
-  {
-    wsrep_dump_rbr_buf_with_header(thd, data.data(), data.size());
-    thd->wsrep_has_ignored_error= false;
-    /* todo: error voting */
-  }
+  int ret= apply_events(thd, m_rli, data, err);
+  thd->wsrep_has_ignored_error= false;
```
[the remark this answers is not in the document]
Comment (ayurchen) [reply sql/wsrep_high_priority_service.cc]: fixed

**Comment by LenzGr** [comment]
Does this PR resolve https://github.com/codership/galera/issues/2 by any chance?

**Pull request merged**